A_KEV precision fails unit test

There's a problem using more precision for the A_KEV constant.  The number coded is one digit less than the 2018 CODATA value. The 4 on the end causes the unit test to
fail when trying the 100,1,1 reflection. 

On Thu, May 20, 2021, 6:29 PM Maksim Rakitin ***@***.***>
wrote:

> There it is (with different units):
>
> quantity
> value
>
> inverse meter-electron volt relationship
> 1.239841984e-06 eV
>
> Yes, we can use that too:
>
> In [10]: constants.physical_constants['inverse meter-electron volt relationship']Out[10]: (1.239841984e-06, 'eV', 0.0)
> In [11]: constants.physical_constants['inverse meter-electron volt relationship'][0]Out[11]: 1.239841984e-06
> In [12]: constants.physical_constants['inverse meter-electron volt relationship'][0] * 1e10Out[12]: 12398.41984
> In [13]: constants.physical_constants['inverse meter-electron volt relationship'][0] * 1e10 / 1e3Out[13]: 12.39841984
>
> But hard-coding it is probably less error-prone.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/bluesky/hklpy/pull/153#issuecomment-845548341>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARMUMCVUQZ7RR7HVEUUE23TOWLNTANCNFSM445A2SIQ>
> .
>

_Originally posted by @prjemian in https://github.com/bluesky/hklpy/issues/153#issuecomment-845564569_

Now: A_KEV = 12.3984198  # 1 Angstrom ~= 12.39842 keV

Switching to:

```python
A_KEV = 12.39841984  # 1 Angstrom ~= 12.39842 keV
```

produces this pytest report:

<details>

```bash
(bluesky_2021_2) prjemian@zap:~/.../Bluesky/hklpy$ pytest -vvv
================================================================ test session starts =================================================================
platform linux -- Python 3.8.10, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- /home/prjemian/.conda/envs/bluesky_2021_2/bin/python
cachedir: .pytest_cache
rootdir: /home/prjemian/Documents/projects/Bluesky/hklpy, configfile: pytest.ini
plugins: anyio-3.3.0
collected 138 items                                                                                                                                  

hkl/tests/test_diffract.py::test_calc_energy_permit PASSED                                                                                     [  0%]
hkl/tests/test_diffract.py::test_energy PASSED                                                                                                 [  1%]
hkl/tests/test_diffract.py::test_energy_offset PASSED                                                                                          [  2%]
hkl/tests/test_diffract.py::test_energy_offset_units PASSED                                                                                    [  2%]
hkl/tests/test_diffract.py::test_energy_units_931eV PASSED                                                                                     [  3%]
hkl/tests/test_diffract.py::test_energy_units_issue79 PASSED                                                                                   [  4%]
hkl/tests/test_diffract.py::test_energy_units_offset PASSED                                                                                    [  5%]
hkl/tests/test_diffract.py::test_energy_units_issue86 PASSED                                                                                   [  5%]
hkl/tests/test_diffract.py::test_names PASSED                                                                                                  [  6%]
hkl/tests/test_diffract.py::test_forward_solutions_table FAILED                                                                                [  7%]
hkl/tests/test_diffract.py::test_pa PASSED                                                                                                     [  7%]
hkl/tests/test_diffract.py::test_wh PASSED                                                                                                     [  8%]
hkl/tests/test_diffract.py::test_show_constraints PASSED                                                                                       [  9%]
hkl/tests/test_diffract.py::test_apply_constraints PASSED                                                                                      [ 10%]
hkl/tests/test_diffract.py::test_specify_engine PASSED                                                                                         [ 10%]
hkl/tests/test_extra_motor.py::test_plain_fourc_not_fatal PASSED                                                                               [ 11%]
hkl/tests/test_extra_motor.py::test_extra_real_but_not_fatal PASSED                                                                            [ 12%]
hkl/tests/test_extra_motor.py::test_extra_real_not_fatal PASSED                                                                                [ 13%]
hkl/tests/test_extra_motor.py::test_extra_pseudo_not_fatal PASSED                                                                              [ 13%]
hkl/tests/test_extra_motor.py::test_fourc_position PASSED                                                                                      [ 14%]
hkl/tests/test_extra_motor.py::test_fourc_extra_pseudo PASSED                                                                                  [ 15%]
hkl/tests/test_fourc.py::test_initial PASSED                                                                                                   [ 15%]
hkl/tests/test_fourc.py::test_limits PASSED                                                                                                    [ 16%]
hkl/tests/test_fourc.py::test_check_value PASSED                                                                                               [ 17%]
hkl/tests/test_fourc.py::test_move PASSED                                                                                                      [ 18%]
hkl/tests/test_fourc.py::test_hl_scan PASSED                                                                                                   [ 18%]
hkl/tests/test_fourc.py::test_h00_scan PASSED                                                                                                  [ 19%]
hkl/tests/test_fourc.py::test_hkl_scan PASSED                                                                                                  [ 20%]
hkl/tests/test_fourc.py::test_hkl_range_error PASSED                                                                                           [ 21%]
hkl/tests/test_fourc.py::test_real_axis PASSED                                                                                                 [ 21%]
hkl/tests/test_fourc.py::test_axis_contention PASSED                                                                                           [ 22%]
hkl/tests/test_fourc.py::test_real_axis_range_multi PASSED                                                                                     [ 23%]
hkl/tests/test_fourc.py::test_real_axis_range_error PASSED                                                                                     [ 23%]
hkl/tests/test_init.py::test_shortcuts[calc-A_KEV] PASSED                                                                                      [ 24%]
hkl/tests/test_init.py::test_shortcuts[calc-UnreachableError] PASSED                                                                           [ 25%]
hkl/tests/test_init.py::test_shortcuts[geometries-E4CH] PASSED                                                                                 [ 26%]
hkl/tests/test_init.py::test_shortcuts[geometries-E4CV] PASSED                                                                                 [ 26%]
hkl/tests/test_init.py::test_shortcuts[geometries-E6C] PASSED                                                                                  [ 27%]
hkl/tests/test_init.py::test_shortcuts[geometries-K4CV] PASSED                                                                                 [ 28%]
hkl/tests/test_init.py::test_shortcuts[geometries-K6C] PASSED                                                                                  [ 28%]
hkl/tests/test_init.py::test_shortcuts[geometries-Petra3_p09_eh2] PASSED                                                                       [ 29%]
hkl/tests/test_init.py::test_shortcuts[geometries-SimMixin] PASSED                                                                             [ 30%]
hkl/tests/test_init.py::test_shortcuts[geometries-SimulatedE4CV] PASSED                                                                        [ 31%]
hkl/tests/test_init.py::test_shortcuts[geometries-SimulatedE6C] PASSED                                                                         [ 31%]
hkl/tests/test_init.py::test_shortcuts[geometries-SimulatedK4CV] PASSED                                                                        [ 32%]
hkl/tests/test_init.py::test_shortcuts[geometries-SimulatedK6C] PASSED                                                                         [ 33%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilMars] PASSED                                                                           [ 34%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilSiriusKappa] PASSED                                                                    [ 34%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilSiriusTurret] PASSED                                                                   [ 35%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilSixsMed1p2] PASSED                                                                     [ 36%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilSixsMed2p2] PASSED                                                                     [ 36%]
hkl/tests/test_init.py::test_shortcuts[geometries-SoleilSixsMed2p3] PASSED                                                                     [ 37%]
hkl/tests/test_init.py::test_shortcuts[geometries-Zaxis] PASSED                                                                                [ 38%]
hkl/tests/test_init.py::test_shortcuts[user-cahkl_table] PASSED                                                                                [ 39%]
hkl/tests/test_init.py::test_shortcuts[user-cahkl] PASSED                                                                                      [ 39%]
hkl/tests/test_init.py::test_shortcuts[user-calc_UB] PASSED                                                                                    [ 40%]
hkl/tests/test_init.py::test_shortcuts[user-change_sample] PASSED                                                                              [ 41%]
hkl/tests/test_init.py::test_shortcuts[user-list_samples] PASSED                                                                               [ 42%]
hkl/tests/test_init.py::test_shortcuts[user-new_sample] PASSED                                                                                 [ 42%]
hkl/tests/test_init.py::test_shortcuts[user-or_swap] PASSED                                                                                    [ 43%]
hkl/tests/test_init.py::test_shortcuts[user-pa] PASSED                                                                                         [ 44%]
hkl/tests/test_init.py::test_shortcuts[user-select_diffractometer] PASSED                                                                      [ 44%]
hkl/tests/test_init.py::test_shortcuts[user-set_energy] PASSED                                                                                 [ 45%]
hkl/tests/test_init.py::test_shortcuts[user-setor] PASSED                                                                                      [ 46%]
hkl/tests/test_init.py::test_shortcuts[user-show_sample] PASSED                                                                                [ 47%]
hkl/tests/test_init.py::test_shortcuts[user-show_selected_diffractometer] PASSED                                                               [ 47%]
hkl/tests/test_init.py::test_shortcuts[user-update_sample] PASSED                                                                              [ 48%]
hkl/tests/test_init.py::test_shortcuts[user-wh] PASSED                                                                                         [ 49%]
hkl/tests/test_init.py::test_shortcuts[util-Constraint] PASSED                                                                                 [ 50%]
hkl/tests/test_init.py::test_shortcuts[util-diffractometer_types] PASSED                                                                       [ 50%]
hkl/tests/test_init.py::test_shortcuts[util-get_position_tuple] PASSED                                                                         [ 51%]
hkl/tests/test_init.py::test_shortcuts[util-Lattice] PASSED                                                                                    [ 52%]
hkl/tests/test_init.py::test_shortcuts[util-list_orientation_runs0] PASSED                                                                     [ 52%]
hkl/tests/test_init.py::test_shortcuts[util-list_orientation_runs1] PASSED                                                                     [ 53%]
hkl/tests/test_init.py::test_shortcuts[util-restore_constraints] PASSED                                                                        [ 54%]
hkl/tests/test_init.py::test_shortcuts[util-restore_energy] PASSED                                                                             [ 55%]
hkl/tests/test_init.py::test_shortcuts[util-restore_orientation] PASSED                                                                        [ 55%]
hkl/tests/test_init.py::test_shortcuts[util-restore_reflections0] PASSED                                                                       [ 56%]
hkl/tests/test_init.py::test_shortcuts[util-restore_reflections1] PASSED                                                                       [ 57%]
hkl/tests/test_init.py::test_shortcuts[util-restore_UB] PASSED                                                                                 [ 57%]
hkl/tests/test_init.py::test_shortcuts[util-run_orientation_info] PASSED                                                                       [ 58%]
hkl/tests/test_init.py::test_shortcuts[util-software_versions] PASSED                                                                          [ 59%]
hkl/tests/test_sample.py::test_compute_UB PASSED                                                                                               [ 60%]
hkl/tests/test_sample.py::test_orientation_reflections PASSED                                                                                  [ 60%]
hkl/tests/test_sample.py::test_reflections_details PASSED                                                                                      [ 61%]
hkl/tests/test_sample.py::test_swap_orientation_reflections PASSED                                                                             [ 62%]
hkl/tests/test_save_restore_UB.py::test_basic_setup FAILED                                                                                     [ 63%]
hkl/tests/test_save_restore_UB.py::test_fourc_orientation_save PASSED                                                                          [ 63%]
hkl/tests/test_save_restore_UB.py::test_fourc_run_orientation_info FAILED                                                                      [ 64%]
hkl/tests/test_save_restore_UB.py::test_list_orientation_runs PASSED                                                                           [ 65%]
hkl/tests/test_save_restore_UB.py::test_restore_orientation PASSED                                                                             [ 65%]
hkl/tests/test_save_restore_UB.py::test_restore_sample PASSED                                                                                  [ 66%]
hkl/tests/test_tardis.py::test_params PASSED                                                                                                   [ 67%]
hkl/tests/test_tardis.py::test_reachable PASSED                                                                                                [ 68%]
hkl/tests/test_tardis.py::test_inversion PASSED                                                                                                [ 68%]
hkl/tests/test_tardis.py::test_unreachable PASSED                                                                                              [ 69%]
hkl/tests/test_tardis.py::test_issue62[0.0-0.0-0.0-0.0-0.0-0.0] PASSED                                                                         [ 70%]
hkl/tests/test_tardis.py::test_issue62[0.075-0.125-0.0-5.3e-05-8.3e-05-0.00078] PASSED                                                         [ 71%]
hkl/tests/test_tardis.py::test_issue62[0.15-0.25-0.0-0.000107-0.000166-0.001561] PASSED                                                        [ 71%]
hkl/tests/test_tardis.py::test_issue62[0.225-0.375-0.0-0.00016-0.000249-0.002341] PASSED                                                       [ 72%]
hkl/tests/test_tardis.py::test_issue62[0.3-0.5-0.0-0.000214-0.000333-0.003121] PASSED                                                          [ 73%]
hkl/tests/test_tardis.py::test_sample1 PASSED                                                                                                  [ 73%]
hkl/tests/test_tardis.py::test_sample1_calc_only PASSED                                                                                        [ 74%]
hkl/tests/test_tardis.py::test_sample1_UB PASSED                                                                                               [ 75%]
hkl/tests/test_tardis.py::test_tardis_forward[1.61198-4-4-0-angles0] PASSED                                                                    [ 76%]
hkl/tests/test_tardis.py::test_tardis_forward[1.61198-4-1-0-angles1] PASSED                                                                    [ 76%]
hkl/tests/test_tardis.py::test_tardis_forward[1.60911-6-0-0-angles2] PASSED                                                                    [ 77%]
hkl/tests/test_tardis.py::test_tardis_forward[1.60954-3-2-0-angles3] PASSED                                                                    [ 78%]
hkl/tests/test_tardis.py::test_tardis_forward[1.60954-5-4-0-angles4] PASSED                                                                    [ 78%]
hkl/tests/test_tardis.py::test_tardis_forward[1.60954-4-5-0-angles5] PASSED                                                                    [ 79%]
hkl/tests/test_user.py::test_select_diffractometer PASSED                                                                                      [ 80%]
hkl/tests/test_user.py::test_cahkl PASSED                                                                                                      [ 81%]
hkl/tests/test_user.py::test_cahkl_table PASSED                                                                                                [ 81%]
hkl/tests/test_user.py::test_calc_UB PASSED                                                                                                    [ 82%]
hkl/tests/test_user.py::test_list_samples PASSED                                                                                               [ 83%]
hkl/tests/test_user.py::test_new_sample PASSED                                                                                                 [ 84%]
hkl/tests/test_user.py::test_set_energy PASSED                                                                                                 [ 84%]
hkl/tests/test_user.py::test_setor PASSED                                                                                                      [ 85%]
hkl/tests/test_user.py::test_show_sample PASSED                                                                                                [ 86%]
hkl/tests/test_user.py::test_update_sample PASSED                                                                                              [ 86%]
hkl/tests/test_user.py::test_pa PASSED                                                                                                         [ 87%]
hkl/tests/test_user.py::test_wh PASSED                                                                                                         [ 88%]
hkl/tests/test_util.py::test__package_info_states PASSED                                                                                       [ 89%]
hkl/tests/test_util.py::test_get_package_info[bluesky-1.6] PASSED                                                                              [ 89%]
hkl/tests/test_util.py::test_get_package_info[gobject-introspection-1.68.0] PASSED                                                             [ 90%]
hkl/tests/test_util.py::test_get_package_info[hkl-5.0.0] PASSED                                                                                [ 91%]
hkl/tests/test_util.py::test_get_package_info[hklpy-0] PASSED                                                                                  [ 92%]
hkl/tests/test_util.py::test_get_package_info[no-such-package----] PASSED                                                                      [ 92%]
hkl/tests/test_util.py::test_get_package_info[ophyd-1.6] PASSED                                                                                [ 93%]
hkl/tests/test_util.py::test_software_versions_default_list[None] PASSED                                                                       [ 94%]
hkl/tests/test_util.py::test_software_versions_default_list[case1] PASSED                                                                      [ 94%]
hkl/tests/test_util.py::test_software_versions_default_list[] PASSED                                                                           [ 95%]
hkl/tests/test_util.py::test_software_versions_items[bluesky-1.6] PASSED                                                                       [ 96%]
hkl/tests/test_util.py::test_software_versions_items[gobject-introspection-1.68.0] PASSED                                                      [ 97%]
hkl/tests/test_util.py::test_software_versions_items[hkl-5.0.0] PASSED                                                                         [ 97%]
hkl/tests/test_util.py::test_software_versions_items[hklpy-0] PASSED                                                                           [ 98%]
hkl/tests/test_util.py::test_software_versions_items[no-such-package----] PASSED                                                               [ 99%]
hkl/tests/test_util.py::test_software_versions_items[ophyd-1.6] PASSED                                                                         [100%]

====================================================================== FAILURES ======================================================================
____________________________________________________________ test_forward_solutions_table ____________________________________________________________

fourc = Fourc(prefix='', name='fourc', settle_time=0.0, timeout=None, egu='', limits=(0, 0), source='computed', read_attrs=['h... '_hklpy_version', '_pseudos', '_reals', '_constraints', '_mode', 'orientation_attrs', 'h', 'k', 'l'], concurrent=True)

    def test_forward_solutions_table(fourc):
        fourc.energy.put(A_KEV / 1.54)
    
        # (100) has chi ~ 0 which poses occasional roundoff errors
        # (sometimes -0.00000, sometimes 0.00000)
        sol = fourc.forward(1, 0, 0)
        assert pytest.approx(sol.omega, 1e-5) == -30
        assert pytest.approx(sol.chi, 1e-5) == 0
        assert pytest.approx(sol.phi, 1e-5) == -90
        assert pytest.approx(sol.tth, 1e-5) == -60
    
        tbl = fourc.forward_solutions_table(
            # fmt: off
            [
                [1, 1, 0],
                [1, 1, 1],
                [100, 1, 1],  # no solutions
            ]
            # fmt: on
        )
        received = str(tbl).splitlines()
        expected = [
            "=========== ======== ===== ======== ==== =====",
            "(hkl)       solution omega chi      phi  tth  ",
            "=========== ======== ===== ======== ==== =====",
            "[1, 1, 0]   0        45.0  45.0     90.0 90.0 ",
            "[1, 1, 1]   0        60.0  35.26439 45.0 120.0",
            "[100, 1, 1] none                              ",
            "=========== ======== ===== ======== ==== =====",
        ]
        for r, e in zip(received, expected):
>           assert r == e
E           AssertionError: assert '=========== ======== ===== ======== ===== =====' == '=========== ======== ===== ======== ==== ====='
E             - =========== ======== ===== ======== ==== =====
E             ?                                     -----
E             + =========== ======== ===== ======== ===== =====
E             ?                                          ++++++

hkl/tests/test_diffract.py:196: AssertionError
__________________________________________________________________ test_basic_setup __________________________________________________________________

fourc = Fourc(prefix='', name='fourc', settle_time=0.0, timeout=None, egu='', limits=(0, 0), source='computed', read_attrs=['h... '_hklpy_version', '_pseudos', '_reals', '_constraints', '_mode', 'orientation_attrs', 'h', 'k', 'l'], concurrent=True)
kappa = Kappa(prefix='', name='kappa', settle_time=0.0, timeout=None, egu='', limits=(0, 0), source='computed', read_attrs=['h... '_hklpy_version', '_pseudos', '_reals', '_constraints', '_mode', 'orientation_attrs', 'h', 'k', 'l'], concurrent=True)

    def test_basic_setup(fourc, kappa):
        assert fourc is not None
        assert kappa is not None
        assert fourc != kappa
        assert fourc._reals != kappa._reals
    
>       assert fourc.calc.wavelength == 1.54
E       assert 1.5399999999999998 == 1.54
E         +1.5399999999999998
E         -1.54

hkl/tests/test_save_restore_UB.py:79: AssertionError
__________________________________________________________ test_fourc_run_orientation_info ___________________________________________________________

cat = <Intake catalog: temp>, RE = <bluesky.run_engine.RunEngine object at 0x7f515adc9c70>
fourc = Fourc(prefix='', name='fourc', settle_time=0.0, timeout=None, egu='', limits=(0, 0), source='computed', read_attrs=['h... '_hklpy_version', '_pseudos', '_reals', '_constraints', '_mode', 'orientation_attrs', 'h', 'k', 'l'], concurrent=True)

    def test_fourc_run_orientation_info(cat, RE, fourc):
        RE(bp.count([fourc]))
        info = hkl.util.run_orientation_info(cat[1])
        assert info is not None
        assert isinstance(info, dict)
        assert "fourc" in info
        fourc_orient = info["fourc"]
        assert "orientation_attrs" in fourc_orient
        assert "reflections_details" in fourc_orient["orientation_attrs"]
        assert "reflections_details" in fourc_orient
        refls = fourc_orient["reflections_details"]
        assert len(refls) == 2
        assert "wavelength" in refls[0]
>       assert refls[0]["wavelength"] == 1.54
E       assert 1.5399999999999998 == 1.54
E         +1.5399999999999998
E         -1.54

hkl/tests/test_save_restore_UB.py:143: AssertionError
================================================================== warnings summary ==================================================================
../../../../.local/lib/python3.8/site-packages/pims/cine.py:29
  /home/prjemian/.local/lib/python3.8/site-packages/pims/cine.py:29: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    from collections import Iterable

hkl/tests/test_save_restore_UB.py: 36 warnings
  /home/prjemian/.conda/envs/bluesky_2021_2/lib/python3.8/site-packages/databroker/v1.py:961: PendingDeprecationWarning: The method Broker.insert may be removed in a future release of databroker.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================== short test summary info ===============================================================
FAILED hkl/tests/test_diffract.py::test_forward_solutions_table - AssertionError: assert '=========== ======== ===== ======== ===== =====' == '====...
FAILED hkl/tests/test_save_restore_UB.py::test_basic_setup - assert 1.5399999999999998 == 1.54
FAILED hkl/tests/test_save_restore_UB.py::test_fourc_run_orientation_info - assert 1.5399999999999998 == 1.54
==================================================== 3 failed, 135 passed, 37 warnings in 14.57s =====================================================
```

</details>

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -11,6 +11,17 @@ Release History
     Deprecations
     Contributors
 
+v1.0.1 (2021-09-20 or thereabouts)
+======================
+
+Maintenance release.
+
+Maintenance
+-----------
+
+* Increase precision of ``A_KEV`` constant to match 2018 CODATA value.
+* Provide project description for PyPI packaging.
+
 v1.0.0 (2021-09-13)
 ======================
 

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -12,7 +12,7 @@ Release History
     Contributors
 
 v1.0.1 (2021-09-20 or thereabouts)
-======================
+==================================
 
 Maintenance release.
 

--- a/hkl/calc.py
+++ b/hkl/calc.py
@@ -64,7 +64,7 @@ logger = logging.getLogger(__name__)
 
 # per NIST publication of CODATA Fundamental Physical Constants
 # https://www.nist.gov/programs-projects/codata-values-fundamental-physical-constants
-A_KEV = 12.3984198  # 1 Angstrom ~= 12.39842 keV
+A_KEV = 12.39841984  # 1 Angstrom ~= 12.39842 keV
 # A_KEV = 12.39842  # 1 Angstrom ~= 12.39842 keV
 NM_KEV = A_KEV * 0.1  # 1 nm ~= 1.239842 keV
 # Note: NM_KEV is not used by hklpy now, remains here as legacy

--- a/hkl/calc.py
+++ b/hkl/calc.py
@@ -65,7 +65,6 @@ logger = logging.getLogger(__name__)
 # per NIST publication of CODATA Fundamental Physical Constants
 # https://www.nist.gov/programs-projects/codata-values-fundamental-physical-constants
 A_KEV = 12.39841984  # 1 Angstrom ~= 12.39842 keV
-# A_KEV = 12.39842  # 1 Angstrom ~= 12.39842 keV
 NM_KEV = A_KEV * 0.1  # 1 nm ~= 1.239842 keV
 # Note: NM_KEV is not used by hklpy now, remains here as legacy
 

--- a/hkl/tests/test_diffract.py
+++ b/hkl/tests/test_diffract.py
@@ -173,6 +173,12 @@ def test_forward_solutions_table(fourc):
     assert pytest.approx(sol.phi, 1e-5) == -90
     assert pytest.approx(sol.tth, 1e-5) == -60
 
+    fourc.apply_constraints(
+        {
+            "tth": Constraint(0, 180, 0, True),
+        }
+    )
+
     tbl = fourc.forward_solutions_table(
         # fmt: off
         [

--- a/hkl/tests/test_save_restore_UB.py
+++ b/hkl/tests/test_save_restore_UB.py
@@ -76,7 +76,7 @@ def test_basic_setup(fourc, kappa):
     assert fourc != kappa
     assert fourc._reals != kappa._reals
 
-    assert fourc.calc.wavelength == 1.54
+    assert round(fourc.calc.wavelength, 2) == 1.54
     assert fourc.calc.wavelength == kappa.calc.wavelength
 
 
@@ -140,7 +140,7 @@ def test_fourc_run_orientation_info(cat, RE, fourc):
     refls = fourc_orient["reflections_details"]
     assert len(refls) == 2
     assert "wavelength" in refls[0]
-    assert refls[0]["wavelength"] == 1.54
+    assert round(refls[0]["wavelength"], 2) == 1.54
 
 
 def test_list_orientation_runs(cat, RE, fourc, kappa):

